dlt/destinations/utils.py

```diff
@@ -77,17 +77,32 @@ def verify_sql_job_client_schema(schema: Schema, warnings: bool = True) -> List[
                         f"""Allowed values: {', '.join(['"' + s + '"' for s in MERGE_STRATEGIES])}.""",
                     )
                 )
-            if (
-                table.get("x-merge-strategy") == "delete-insert"
-                and not has_column_with_prop(table, "primary_key")
-                and not has_column_with_prop(table, "merge_key")
-            ):
-                log(
-                    f"Table {table_name} has `write_disposition` set to `merge`"
-                    " and `merge_strategy` set to `delete-insert`, but no primary or"
-                    " merge keys defined."
-                    " dlt will fall back to `append` for this table."
-                )
+            if table.get("x-merge-strategy") == "delete-insert":
+                if not has_column_with_prop(table, "primary_key") and not has_column_with_prop(
+                    table, "merge_key"
+                ):
+                    log(
+                        f"Table {table_name} has `write_disposition` set to `merge`"
+                        " and `merge_strategy` set to `delete-insert`, but no primary or"
+                        " merge keys defined."
+                        " dlt will fall back to `append` for this table."
+                    )
+            elif table.get("x-merge-strategy") == "upsert":
+                if not has_column_with_prop(table, "primary_key"):
+                    exception_log.append(
+                        SchemaCorruptedException(
+                            schema.name,
+                            f"No primary key defined for table `{table['name']}`."
+                            " `primary_key` needs to be set when using the `upsert`"
+                            " merge strategy.",
+                        )
+                    )
+                if has_column_with_prop(table, "merge_key"):
```

Comment:
I noticed these warnings are logged more than once because `verify_sql_job_client_schema` is called multiple times in a single pipeline run. Do we have something to prevent that?

Reply:
I made a commit to implement this but [reverted](https://github.com/dlt-hub/dlt/pull/1466/commits/cec667170817860bd4c1b050710314d471ba1d79) it because some tests were failing and we have a bit of urgency to merge this PR. Let's do it in another PR.